src/localization/gps_linearization.py

```diff
@@ -28,59 +31,90 @@ class GPSLinearization:
-        self.odom_frame = rospy.get_param("gps_linearization/odom_frame")
-        self.rover_frame = rospy.get_param("gps_linearization/rover_frame")
+        self.publish_tf = rospy.get_param("gps_linearization/publish_tf")
+        self.use_odom = rospy.get_param("use_odom_frame")
+
+        self.world_frame = rospy.get_param("world_frame")
+        self.odom_frame = rospy.get_param("odom_frame")
+        self.rover_frame = rospy.get_param("rover_frame")
+
+        self.use_dop_cov = rospy.get_param("global_ekf/use_gps_dop_covariance")
+        self.config_gps_covariance = np.array(rospy.get_param("global_ekf/gps_covariance", None))
+
+        # init to zero pose and either zero covariance or configured covariance
+        self.pose = SE3()
+        self.covariance = np.zeros((6, 6))
+        if not self.use_dop_cov:
+            self.covariance[:3, :3] = self.config_gps_covariance
+
+        rospy.Subscriber("gps/fix", NavSatFix, self.gps_callback)
+        rospy.Subscriber("imu/data", ImuAndMag, self.imu_callback)
+        self.pose_publisher = rospy.Publisher("gps/pose", PoseWithCovarianceStamped, queue_size=1)
+
+        if self.publish_tf:
+            self.tf_broadcaster = tf2_ros.TransformBroadcaster()
+            self.tf_buffer = tf2_ros.Buffer()
+            self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
 
     def publish_pose(self):
         """
-        Publishes the pose of the rover in relative to the map frame to the TF tree,
-        either as a direct map->base_link transform, or if the odom frame is in use,
-        indirectly as a map->odom transform.
+        Publishes the pose of the rover relative to the map frame, either to the TF tree or as
+        a PoseWithCovarianceStamped message. The pose will be published either as a direct
+        map->base_link transform, or an indirect map-odom transform if the odom frame is in use.
+        See the wiki for more details:
+        https://github.com/umrover/mrover-ros/wiki/Localization#guide-to-localization-frames
         """
         rover_in_map = self.pose
 
-        if self.use_odom:
-            # Get the odom to rover transform from the TF tree
-            rover_in_odom = SE3.from_tf_tree(self.tf_buffer, self.odom_frame, self.rover_frame)
-            odom_to_rover = rover_in_odom.transform_matrix()
-            map_to_rover = rover_in_map.transform_matrix()
-
-            # Calculate the intermediate transform from the overall transform and odom to rover
-            map_to_odom = map_to_rover @ np.linalg.inv(odom_to_rover)
-            odom_in_map = SE3.from_transform_matrix(map_to_odom)
-            odom_in_map.publish_to_tf_tree(
-                self.tf_broadcaster, parent_frame=self.world_frame, child_frame=self.odom_frame
-            )
+        if self.publish_tf:
+            if self.use_odom:
+                # TODO: fix naming conventions of transforms per Ashwins instructions
```

Comment:
could you expand on this?

Reply:
oh yeah I still need to do that, basically some of the transforms are flipped, so they should be called a to b instead of b to a

Reply:
done, @niwhsa9 does this look correct now? lines 90-96 of gps_linearization

Reply:
hm, it still doesn't seem correct. Let me illustrate with one of the variables. `rover_in_odom` is equivalent to `rover_to_odom` which is the reverse of what you have written here.

Reply:
I think you somehow managed to comment on an outdated version, the stuff I pushed does what you say: `rover_to_odom = rover_in_odom.transform_matrix()`

Reply:
yes this is now correct